src/oscar/apps/address/abstract_models.py

```diff
@@ -351,6 +344,36 @@ def join_fields(self, fields, separator=u", "):
             else:
                 value = getattr(self, field)
             field_values.append(value)
+        return field_values
+
+    def get_address_field_values(self, fields, include_salutation=True, include_country=True):
+        """
+        Returns set of field values within the salutation and country.
+        """
+        field_values = self.get_field_values(fields)
+        if include_salutation:
+            field_values = [self.salutation] + self.get_field_values(fields)
+        field_values = [f.strip() for f in field_values if f]
+        if include_country:
+            try:
+                field_values.append(self.country.printable_name)
+            except exceptions.ObjectDoesNotExist:
+                pass
+        return field_values
+
+    def generate_hash(self):
+        """
+        Returns a hash of the address, based on standard set of fields, listed
+        out in `hash_fields` property.
+        """
+        field_values = self.get_address_field_values(self.hash_fields)
```

Comment:
Need a test that checks that a hash is correctly generated when `hash_fields` is overridden (but `base_fields` is not).

Reply:
Ok done.